bridges/snowbridge/primitives/outbound-queue/src/v2/converter/mod.rs

```diff
@@ -0,0 +1,216 @@
+
+		if network != expected_network {
+			log::trace!(target: TARGET, "skipped due to unmatched bridge network {network:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		// Cloning destination to avoid modifying the value so subsequent exporters can use it.
+		let dest = destination.clone().ok_or(SendError::MissingArgument)?;
+		if dest != Here {
+			log::trace!(target: TARGET, "skipped due to unmatched remote destination {dest:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		// Cloning universal_source to avoid modifying the value so subsequent exporters can use it.
+		let (local_net, local_sub) = universal_source.clone()
+            .ok_or_else(|| {
+                log::error!(target: TARGET, "universal source not provided.");
+                SendError::MissingArgument
+            })?
+            .split_global()
+            .map_err(|()| {
+                log::error!(target: TARGET, "could not get global consensus from universal source '{universal_source:?}'.");
+                SendError::NotApplicable
+            })?;
+
+		if Ok(local_net) != universal_location.global_consensus() {
+			log::trace!(target: TARGET, "skipped due to unmatched relay network {local_net:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		let para_id = match local_sub.as_slice() {
+			[Parachain(para_id)] => *para_id,
+			_ => {
+				log::error!(target: TARGET, "could not get parachain id from universal source '{local_sub:?}'.");
+				return Err(SendError::NotApplicable)
+			},
+		};
+
+		if ParaId::from(para_id) != AssetHubParaId::get() {
+			log::error!(target: TARGET, "is not from asset hub '{para_id:?}'.");
+			return Err(SendError::NotApplicable)
+		}
+
+		let message = message.clone().ok_or_else(|| {
+			log::error!(target: TARGET, "xcm message not provided.");
+			SendError::MissingArgument
+		})?;
+
+		// Inspect `AliasOrigin` as V2 message. This exporter should only process Snowbridge V2
+		// messages. We use the presence of an `AliasOrigin` instruction to distinguish between
+		// Snowbridge V2 and Snowbridge V1 messages, since XCM V5 came after Snowbridge V1 and
+		// so is not supported in Snowbridge V1. Snowbridge V1 messages are processed by the
+		// snowbridge-outbound-queue-primitives v1 exporter.
+		let mut instructions = message.clone().0;
+		let result = instructions.matcher().match_next_inst_while(
+			|_| true,
+			|inst| {
+				return match inst {
+					AliasOrigin(..) => Err(ProcessMessageError::Yield),
+					_ => Ok(ControlFlow::Continue(())),
+				}
+			},
+		);
+		ensure!(result.is_err(), SendError::NotApplicable);
+
+		let mut converter = XcmConverter::<ConvertAssetId, ()>::new(&message, expected_network);
+		let message = converter.convert().map_err(|err| {
+			log::error!(target: TARGET, "unroutable due to pattern matching error '{err:?}'.");
+			SendError::Unroutable
+		})?;
+
+		// validate the message
+		let (ticket, _) = OutboundQueue::validate(&message).map_err(|err| {
```

Comment:
Why is there no AH->BH XCM transport fee? If this is by design, it should be documented/explained somewhere around here.

Reply:
So this code runs on BH. `XcmConverter` converts the XCM to a bridge command.

We copied the XCM `validate` & `deliver` pattern as the interface for _committing_ messages in the outbound-queue on BH. The fee returned (and not handled) here is basically the weight for committing messages to storage on BH.

@yrong remind me why we are not handling this fee?

Reply:
https://github.com/acatangiu/polkadot-sdk/pull/3

Reply:
https://github.com/paritytech/polkadot-sdk/pull/7402/commits/d5657212a0b54d17defb41b6830284b0e80e666c